File: 058MinimumWaitingTime/minimum_waiting_time.rb
# Solution
# O(nlog(n)) time / O(1) space

def minimum_waiting_time(queries)
  queries.sort!
  total_waiting_time = 0

  (0..queries.length).each do |idx|
    duration = queries[idx]
    queries_left = queries.length - (idx + 1)
    total_waiting_time += duration.to_i * queries_left
  end
  total_waiting_time
end

File: 058MinimumWaitingTime/minimumWaitingTime.js
// Solution
// O(nlog(n)) time / O(1) space

const minimumWaitingTime = (queries) => {
  queries.sort((a, b) => a - b);

  let totalWaitingTime = 0;
  for (let idx = 0; idx < queries.length; idx++) {
    const duration = queries[idx];
    const queriesLeft = queries.length - (idx + 1);
    totalWaitingTime += duration * queriesLeft;
  }
  return totalWaitingTime;
};
